File: backend/src/services/authService.js
import bcrypt from "bcryptjs";
import { User, Patient, Doctor, Pharmacy, sequelize } from "../models/index.js";
import { generateToken } from "../middleware/auth.js";
import { Op } from "sequelize";

class AuthService {
  // Unified registration - creates ONLY User record, role-specific records created on verification
  async registerUser({ name, phone, email, password, gender, dateOfBirth, role, preferredLanguage }) {
    const transaction = await sequelize.transaction();
    
    try {
      // Check if user already exists by phone or email
      const existingUser = await User.findOne({
        where: {
          [Op.or]: [
            { phone },
            ...(email ? [{ email }] : [])
          ]
        }
      });

      if (existingUser) {
        throw new Error('User already exists with this phone number or email');
      }

      // Hash password
      const hashedPassword = await bcrypt.hash(password, 12);

      // Create ONLY User record - no role-specific records yet
      const user = await User.create({
        name,
        phone,
        email,
        password: hashedPassword,
        role,
        preferredLanguage,
        gender,
        dateOfBirth,
        isVerified: role === 'patient' // Only patients are auto-verified
      }, { transaction });

      // For patients, create Patient record immediately since no verification needed
      if (role === 'patient') {
        await Patient.create({
          userId: user.id
        }, { transaction });
      }
      // For doctors and pharmacies, role-specific records will be created during verification

      await transaction.commit();

      // Generate token
      const token = generateToken({
        userId: user.id,
        userType: role,
      });

      return {
        token,
        user: {
          id: user.id,
          name: user.name,
          email: user.email,
          phone: user.phone,
          role: user.role,
          preferredLanguage: user.preferredLanguage,
          isVerified: user.isVerified,
          needsVerification: role !== 'patient'
        }
      };
    } catch (error) {
      await transaction.rollback();
      throw error;
    }
  }

  // Unified login - accepts phone/email and password
  async loginUser(identifier, password) {
    // Find user by phone or email
    const user = await User.findOne({
      where: {
        [Op.or]: [
          { phone: identifier },
          { email: identifier }
        ]
      },
      include: [
        { model: Patient, as: 'patient' },
        { model: Doctor, as: 'doctor' },
        { model: Pharmacy, as: 'pharmacy' }
      ]
    });

    if (!user) {
      throw new Error('Invalid credentials');
    }

    // Check password
    const isValidPassword = await bcrypt.compare(password, user.password);
    if (!isValidPassword) {
      throw new Error('Invalid credentials');
    }

    if (!user.isActive) {
      throw new Error('Account is deactivated. Please contact support.');
    }

    // Update last login
    await user.update({ lastLoginAt: new Date() });

    // Generate token
    const token = generateToken({
      userId: user.id,
      userType: user.role,
    });

    // Get role-specific data
    let roleSpecificData = {};
    switch (user.role) {
      case 'doctor':
        if (user.doctor) {
          roleSpecificData = {
            specialization: user.doctor.specialization,
            isVerified: user.doctor.isVerified
          };
        }
        break;
      case 'pharmacy':
        if (user.pharmacy) {
          roleSpecificData = {
            address: user.pharmacy.address,
            isVerified: user.pharmacy.isVerified
          };
        }
        break;
    }

    return {
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        phone: user.phone,
        role: user.role,
        preferredLanguage: user.preferredLanguage,
        isVerified: user.isVerified,
        ...roleSpecificData
      }
    };
  }

  // Doctor verification - creates Doctor record with verification data
  async verifyDoctor(userId, verificationData) {
    const { hospitalId, specialization, experience, languages, currentHospitalClinic, pincode } = verificationData;
    
    const transaction = await sequelize.transaction();
    
    try {
      // Find user
      const user = await User.findByPk(userId);
      
      if (!user || user.role !== 'doctor') {
        throw new Error('Invalid user or user is not a doctor');
      }
      
      if (user.isVerified) {
        throw new Error('Doctor is already verified');
      }
      
      // Check if Doctor record already exists
      let doctorRecord = await Doctor.findOne({ where: { userId } });
      
      if (doctorRecord) {
        // Update existing doctor record
        await doctorRecord.update({
          hospitalId,
          specialization,
          experience,
          languages: Array.isArray(languages) ? languages : [languages],
          currentHospitalClinic,
          pincode,
          isVerified: true,
          verificationData: {
            verifiedAt: new Date(),
            ...verificationData
          }
        }, { transaction });
      } else {
        // Create new doctor record with verification data
        doctorRecord = await Doctor.create({
          userId,
          hospitalId,
          specialization,
          experience,
          languages: Array.isArray(languages) ? languages : [languages],
          currentHospitalClinic,
          pincode,
          isVerified: true,
          verificationData: {
            verifiedAt: new Date(),
            ...verificationData
          }
        }, { transaction });
      }
      
      // Update user verification status
      await user.update({ isVerified: true }, { transaction });
      
      await transaction.commit();
      
      return {
        message: 'Doctor verification completed successfully',
        isVerified: true
      };
    } catch (error) {
      await transaction.rollback();
      throw error;
    }
  }

  // Pharmacy verification - creates Pharmacy record with verification data
  async verifyPharmacy(userId, verificationData) {
    const { storeName, licenseNumber, gstNumber, address, city, state, pincode, workingHours } = verificationData;
    
    const transaction = await sequelize.transaction();
    
    try {
      // Find user
      const user = await User.findByPk(userId);
      
      if (!user || user.role !== 'pharmacy') {
        throw new Error('Invalid user or user is not a pharmacy owner');
      }
      
      if (user.isVerified) {
        throw new Error('Pharmacy is already verified');
      }
      
      // Check if Pharmacy record already exists
      let pharmacyRecord = await Pharmacy.findOne({ where: { userId } });
      
      if (pharmacyRecord) {
        // Update existing pharmacy record
        await pharmacyRecord.update({
          storeName,
          licenseNumber,
          gstNumber,
          address,
          city,
          state,
          pincode,
          workingHours,
          isVerified: true,
          verificationData: {
            verifiedAt: new Date(),
            ...verificationData
          }
        }, { transaction });
      } else {
        // Create new pharmacy record with verification data
        pharmacyRecord = await Pharmacy.create({
          userId,
          storeName,
          licenseNumber,
          gstNumber,
          address,
          city,
          state,
          pincode,
          workingHours,
          isVerified: true,
          verificationData: {
            verifiedAt: new Date(),
            ...verificationData
          }
        }, { transaction });
      }
      
      // Update user verification status
      await user.update({ isVerified: true }, { transaction });
      
      await transaction.commit();
      
      return {
        message: 'Pharmacy verification completed successfully',
        isVerified: true
      };
    } catch (error) {
      await transaction.rollback();
      throw error;
    }
  }

  async refreshToken(refreshToken) {
    // In a real implementation, you would validate the refresh token
    // For now, we'll just return an error
    throw new Error('Refresh token functionality not implemented yet');
  }
}

export const authService = new AuthService();

File: SEHAT-main/sehat_doctor_app/lib/services/auth_service.dart
import 'dart:convert';
import 'base_api_service.dart';
import '../config/api_config.dart';
import '../models/user.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class AuthResponse {
  final String token;
  final User user;

  AuthResponse({required this.token, required this.user});

  factory AuthResponse.fromJson(Map<String, dynamic> json) {
    return AuthResponse(
      token: json['token'] ?? '',
      user: User.fromJson(json['user'] ?? {}),
    );
  }
}

class AuthService {
  static final AuthService _instance = AuthService._internal();
  factory AuthService() => _instance;
  AuthService._internal();

  final BaseApiService _apiService = BaseApiService();
  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage();

  // Register a new user
  Future<AuthResponse> register({
    required String name,
    required String phone,
    required String email,
    required String password,
    required String preferredLanguage,
    required String gender,
    required String dateOfBirth,
    required String role,
  }) async {
    final data = {
      'name': name,
      'phone': phone,
      'email': email,
      'password': password,
      'role': role,
      'preferredLanguage': preferredLanguage,
      'gender': gender,
      'dateOfBirth': dateOfBirth,
    };

    final response = await _apiService.post(ApiConfig.registerEndpoint, data: data);
    
    if (response['status'] == 'success' && response['data'] != null) {
      final authResponse = AuthResponse.fromJson(response['data']);
      await _saveAuthData(authResponse);
      return authResponse;
    } else {
      throw ApiException(response['message'] ?? 'Registration failed');
    }
  }

  // Login with existing credentials
  Future<AuthResponse> login({
    required String identifier,
    required String password,
  }) async {
    final data = {
      'identifier': identifier,
      'password': password,
    };

    final response = await _apiService.post(ApiConfig.loginEndpoint, data: data);
    
    if (response['status'] == 'success' && response['data'] != null) {
      final authResponse = AuthResponse.fromJson(response['data']);
      await _saveAuthData(authResponse);
      return authResponse;
    } else {
      throw ApiException(response['message'] ?? 'Login failed');
    }
  }


  // Verify doctor profile
  Future<void> verifyDoctor({
    required String hospitalId,
    required String specialization,
    required int experience,
    required List<String> languages,
    required String currentHospitalClinic,
    required String pincode,
  }) async {
    final data = {
      'hospitalId': hospitalId,
      'specialization': specialization,
      'experience': experience,
      'languages': languages,
      'currentHospitalClinic': currentHospitalClinic,
      'pincode': pincode,
    };

    final response = await _apiService.post(ApiConfig.verifyDoctorEndpoint, data: data);
    
    if (response['status'] != 'success') {
      throw ApiException(response['message'] ?? 'Verification failed');
    }
  }

  // Verify pharmacy profile
  Future<void> verifyPharmacy({
    required String storeName,
    String? licenseNumber,
    String? gstNumber,
    required String address,
    String? city,
    String? state,
    required String pincode,
    String? workingHours,
  }) async {
    final data = {
      'storeName': storeName,
      if (licenseNumber != null && licenseNumber.isNotEmpty) 'licenseNumber': licenseNumber,
      if (gstNumber != null && gstNumber.isNotEmpty) 'gstNumber': gstNumber,
      'address': address,
      if (city != null && city.isNotEmpty) 'city': city,
      if (state != null && state.isNotEmpty) 'state': state,
      'pincode': pincode,
      if (workingHours != null && workingHours.isNotEmpty) 'workingHours': workingHours,
    };

    final response = await _apiService.post(ApiConfig.verifyPharmacyEndpoint, data: data);
    
    if (response['status'] != 'success') {
      throw ApiException(response['message'] ?? 'Verification failed');
    }
  }

  // Refresh JWT token
  Future<String> refreshToken(String refreshToken) async {
    final data = {'refreshToken': refreshToken};

    final response = await _apiService.post(ApiConfig.refreshTokenEndpoint, data: data);
    
    if (response['status'] == 'success' && response['data'] != null) {
      final token = response['data']['token'];
      await _apiService.setAuthToken(token);
      return token;
    } else {
      throw ApiException(response['message'] ?? 'Token refresh failed');
    }
  }

  // Logout
  Future<void> logout() async {
    try {
      await _apiService.post(ApiConfig.logoutEndpoint);
    } catch (e) {
      // Continue with logout even if API call fails
    }
    
    await _apiService.clearAuthData();
  }

  // Save authentication data to secure storage
  Future<void> _saveAuthData(AuthResponse authResponse) async {
    await _apiService.setAuthToken(authResponse.token);
    await _secureStorage.write(
      key: ApiConfig.userKey, 
      value: jsonEncode(authResponse.user.toJson())
    );
  }

  // Get current user from storage
  Future<User?> getCurrentUser() async {
    try {
      final userData = await _secureStorage.read(key: ApiConfig.userKey);
      if (userData != null) {
        return User.fromJson(jsonDecode(userData));
      }
    } catch (e) {
      print('Error getting current user: $e');
    }
    return null;
  }

  // Check if user is logged in
  Future<bool> isLoggedIn() async {
    final token = await _apiService.getAuthToken();
    return token != null && token.isNotEmpty;
  }

  // Check if doctor needs verification
  Future<bool> needsVerification() async {
    final user = await getCurrentUser();
    return user?.needsVerification ?? false;
  }
}
